### PR TITLE
feat: http server

### LINKED ISSUE #183: feature: provide api for k8sgpt
Checklist:
* [X] I've searched for similar issues and couldn't find anything matching
* [X] I've discussed this feature in the #k8sgpt slack channel

## Is this feature request related to a problem?
* [ ] Yes
* [X] No

## Describe the solution you'd like
Sometimes, it's easier to integrate in other tools via a webservice than a CLI and therefore, it might make sense to provide a simple way to execute k8sgpt via an HTTP request.

Created a small prototype here:

serve

The code is available in the branch feat/serve.

## Benefits for the project and its users

* There would be an option to run k8sgpt in-cluster and let tools consume it
* As a result, integrations for K8sGPT might get very easy to write

## Potential drawbacks
* Security, we might need some authorization
* The k8sgpt binary gets a bit larger (~52M instead of 39M)

### PR DESCRIPTION
**Pull request by thschue**
Closes #183 

## 📑 Description
* Adds `k8sgpt serve` command
* Introduces NewAnalysis function to unify API and cmd operations
* Fixes caching issues with multiple providers

## ✅ Checks

- [X] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [X] All the tests have passed

### CODE REVIEW

File: cmd/serve/serve.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		}
+
+		token := viper.GetString(fmt.Sprintf("%s_key", backendType))
+		// check if nil
+		if token == "" {
+			color.Red("No %s key set. Please run k8sgpt auth", backendType)
+			os.Exit(1)
+		}
+
+		server := server2.K8sGPTServer{
```
Comment (AlexsJones) [inline cmd/serve/serve.go:41]: Could we rename the package import from `server2` to something like `serverProvider` ?

File: pkg/server/main.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+import (
+	"fmt"
+	"github.com/fatih/color"
+	"github.com/k8sgpt-ai/k8sgpt/pkg/analysis"
+	"net/http"
+	"os"
+)
+
+type K8sGPTServer struct {
```
Comment (AlexsJones) [inline pkg/server/main.go:11]: I think renaming this to `Configuration` makes more sense as we don't have a ref to the actual `http.Serve` inside of this object

File: pkg/server/main.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	color.Green("Starting server on port " + s.Port)
+	err := http.ListenAndServe(":"+s.Port, nil)
+	if err != nil {
+		fmt.Printf("error starting server: %s\n", err)
+		return err
+	}
+	return nil
+}
+
+func getBoolParam(param string) bool {
```
Comment (matthisholleville) [inline pkg/server/main.go:69]: Shouldn't we try to convert param to bool and return the result? 

```go
func getBoolParam(param string) bool {
	b, err := strconv.ParseBool(param)
	if err != nil {
		// Handle error if conversion fails
		return false
	}
	return b
}
```

With this modification, even if the user sends "True" the function will interpret the boolean correctly, whereas with the current version, the function will return `false`.

Example:

```go
func getBoolParam(param string) bool {
	b, err := strconv.ParseBool(param)
	if err != nil {
		// Handle error if conversion fails
		return false
	}
	return b
}

func oldGetBoolParam(param string) bool {
	if param == "true" {
		return true
	}
	return false
}

func main() {
	test := "True"
	fmt.Println(getBoolParam(test))
	fmt.Println(oldGetBoolParam(test))
}

---------

go run main.go
true
false
```

File: cmd/serve/serve.go (hunk 1 above)
Comment (thschue) [reply cmd/serve/serve.go]: fixed this

File: pkg/server/main.go (hunk 2 above)
Comment (thschue) [reply pkg/server/main.go]: changed this

File: pkg/server/main.go (hunk 3 above)
Comment (AlexsJones) [reply pkg/server/main.go:69]: +1 on this, 

we could toLower + convert bool

File: pkg/server/main.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		color.Red("Error: %v", err)
+		os.Exit(1)
+	}
+	fmt.Fprintf(w, string(output))
+
+}
+
+func (s *K8sGPTServer) Serve() error {
+	http.HandleFunc("/analyze", s.analyzeHandler)
+	color.Green("Starting server on port " + s.Port)
```
[the remark this answers is not in the document]
Comment (thschue) [reply pkg/server/main.go]: fixed

File: pkg/server/main.go (hunk 3 above)
Comment (thschue) [reply pkg/server/main.go:71]: +1 from my side and fixed

**Review by AlexsJones — APPROVED** [review]

**Pull request merged**